### PR TITLE
Feature/rearrange mc custom types

### PR DESCRIPTION
**Pull request by tcplglo**
Split the set custom types action into two separate actions

### CODE REVIEW

File: cart-service/src/lib/cart-constants.ts [hunk 1]
```diff
@@ -1,8 +1,14 @@
 const fieldPrefix = 'dovetech-discounts-';
 
+export const EXTENSION_TYPES_METADATA_LABEL = `Dovetech Campaigns`;
+export const EXTENSION_TYPES_METADATA_KEY = `${fieldPrefix}extension-metadata`;
+export const DOVETECH_LABEL = `${fieldPrefix}discounts`;
```
Comment (DownChapel) [inline cart-service/src/lib/cart-constants.ts]: I can't see `DOVETECH_LABEL` being used anywhere...

**Review by DownChapel — COMMENTED** [review]
Looks good.

Now the format of the coupon codes has changed, the `commerce-tools-cart-mapper` needs updating to read them in the correct format. This maps them from the commercetools cart to the Dovetech request.

The demo repo will also need to be updated when this has changed. That would be a good way to test out these changes.

File: cart-service/src/lib/cart-constants.ts (hunk 1 above)
Comment (tcplglo) [reply cart-service/src/lib/cart-constants.ts]: leftover removed

**Comment by tcplglo** [comment]
{632A5C89-4894-4845-A994-3CAE56D697E1}
 
tested after changes and seems to apply the discounts correctly

**Comment by tcplglo** [comment]
@DownChapel  this should now be ready for review.
  includes the new custom field definition for data-instance

File: cart-service/src/connector/actions.ts [hunk 2]
```diff
@@ -97,27 +99,43 @@ export async function createCustomTypes(
     .types()
     .get({
       queryArgs: {
-        where: `key = "${CART_METADATA}"`,
+        where: `key = "${EXTENSION_TYPES_DATA_KEY}"`,
       },
     })
     .execute();
 
+  console.log('types', types);
```
Comment (DownChapel) [inline cart-service/src/connector/actions.ts]: Could you remove the console log please?

File: cart-service/src/lib/dovetech-response-mapper.ts [hunk 3]
```diff
@@ -245,21 +244,16 @@ const getDirectDiscountShippingAction = (
...
 ) => {
-  const couponCodes: CouponCode[] = couponCodeAcceptedActions.map((a) => ({
-    code: a.code,
-  }));
-
-  const serialisedCouponCodes = JSON.stringify(couponCodes);
-
-  const fields: { [key: string]: string } = {
+  // set metadata action
+  const fields: { [key: string]: string | string[] } = {
```
Comment (DownChapel) [inline cart-service/src/lib/dovetech-response-mapper.ts]: This looks like the type for the fields:

```suggestion
  const fields: FieldContainer = {
```

File: cart-service/src/lib/dovetech-response-mapper.ts [hunk 4]
```diff
@@ -268,15 +262,18 @@ const buildSetCustomTypeAction = (
...
+      key: EXTENSION_TYPES_DATA_KEY,
       typeId: 'type',
     },
     fields: fields,
   };
-  return setCustomTypeAction;
+
+  const actions: CartUpdateAction[] = [setCustomTypeMetadataAction];
+
+  return actions;
```
Comment (DownChapel) [inline cart-service/src/lib/dovetech-response-mapper.ts]: This function can stay as just returning one action rather than an array now

Comment (tcplglo) [reply cart-service/src/lib/dovetech-response-mapper.ts]: just thought that would leave as an array until we know why they do not let more than type per object at a time. Thought they might give us a reason, or a way to actually use them

File: cart-service/src/connector/actions.ts (hunk 2 above)
Comment (tcplglo) [reply cart-service/src/connector/actions.ts]: removed

File: cart-service/src/lib/dovetech-response-mapper.ts (hunk 3 above)
Comment (tcplglo) [reply cart-service/src/lib/dovetech-response-mapper.ts]: changed

**Review by DownChapel — COMMENTED** [review]
Cool, looks good. Could you just change `CART_EXTENSION_KEY` in `actions.ts` to use dashes too please?

**Review by DownChapel — APPROVED** [review]

**Pull request merged**